src/http/server/handlers/static_file_handler.cr

```diff
@@ -103,6 +94,39 @@ class HTTP::StaticFileHandler
     context.response.headers.add "Location", url
   end
 
+  private def add_cache_headers(response_headers : HTTP::Headers, file_path : String) : Nil
+    last_modified = modification_time(file_path)
+
+    response_headers["Etag"] = etag(last_modified)
+    response_headers["Last-Modified"] = HTTP.format_time(last_modified)
+  end
+
+  private def cache_request?(context : HTTP::Server::Context, file_path : String) : Bool
+    # According to RFC 7232:
+    # A recipient must ignore If-Modified-Since if the request contains an If-None-Match header field
+    if if_none_match = context.request.headers["If-None-Match"]?
+      if_none_match == context.response.headers["Etag"]
+    elsif if_modified_since = context.request.headers["If-Modified-Since"]?
+      header_time = HTTP.parse_time(if_modified_since)
+      # File mtime probably has a higher resolution than the header value.
+      # An exact comparison might be slightly off, so we add 1s padding.
+      # Static files should generally not be modified in subsecond intervals, so this is perfectly safe.
+      # This might be replaced by a more sophisticated time comparison when it becomes available.
+      last_modified = modification_time(file_path)
```

Comment:
Maybe move this to the `call` method and pass `modification_time` to both `add_cache_header` and `cache_request?` to avoid another duplicate file stat.

Reply:
good point! rebased PR once again 😅